### PR TITLE
Repo invitation

### PR DESCRIPTION
**Pull request by danieldevsquad**
I am adding support to manipulate invitations.

I've added an integration test that requires the ability to read a .env file with private authentication info.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,7 @@ phpunit.xml
 composer.lock
 composer.phar
 vendor/*
+#
+.idea
```
Comment (GrahamCampbell) [inline .gitignore]: this should be instead added to your system level gitignore file

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,7 @@ phpunit.xml
...
+.DS_STORE
```
Comment (GrahamCampbell) [inline .gitignore]: same

File: test/Github/Tests/Integration/TestCase.php [hunk 3]
```diff
@@ -33,4 +35,26 @@ public function setUp()
 
         $this->client = $client;
     }
+
+    protected function auth(Client &$client, $accountNumber = 1)
```
Comment (GrahamCampbell) [inline test/Github/Tests/Integration/TestCase.php]: no need for the `&`

File: test/Github/Tests/Integration/TestCase.php [hunk 4]
```diff
@@ -33,4 +35,26 @@ public function setUp()
...
+                case 'token':
+                    $client->authenticate($method, getenv('GITHUB_TOKEN_'.$accountNumber));
+                    break;
+                case 'login':
+                    $client->authenticate($method, getenv('GITHUB_USERNAME_'.$accountNumber),
+                        getenv('GITHUB_PASSWORD_{$accountNumber}'));
+                    break;
+            }
+        } catch (\Exception $e) {
+            error_log('Unable to authenticated', 0);
```
Comment (GrahamCampbell) [inline test/Github/Tests/Integration/TestCase.php]: why not just throw it, i don't think this is useful within a test suite

File: test/Github/Tests/Integration/TestCase.php [hunk 5]
```diff
@@ -36,25 +36,21 @@ public function setUp()
...
-                    $client->authenticate($method, getenv('GITHUB_USERNAME_'.$accountNumber),
-                        getenv('GITHUB_PASSWORD_{$accountNumber}'));
-                    break;
-            }
-        } catch (\Exception $e) {
-            error_log('Unable to authenticated', 0);
+        (new Dotenv(__DIR__.'/../../../../'))->load();
+        $method = getenv('GITHUB_AUTH_METHOD');
+        if (!getenv('GITHUB_AUTH_METHOD')) {
+            return;
```
Comment (m1guelpf) [inline test/Github/Tests/Integration/TestCase.php]: Maybe `$this->markSkipped()` here?

Comment (danieldevsquad) [reply test/Github/Tests/Integration/TestCase.php]: Since the use of env variables only occurs in InvitationsTest.php, I believe it would make sense to move the entire auth logic to it's own setup method, including the markSkipped idea.

File: composer.json [hunk 6]
```diff
@@ -32,7 +32,8 @@
         "php-http/mock-client": "^1.0",
         "guzzlehttp/psr7": "^1.2",
         "sllh/php-cs-fixer-styleci-bridge": "^1.3",
-        "cache/array-adapter": "^0.4"
+        "cache/array-adapter": "^0.4",
+        "vlucas/phpdotenv": "^2.4"
```
Comment (flug) [inline composer.json]: maybe http://symfony.com/doc/current/components/dotenv.html  ??

Comment (danieldevsquad) [reply composer.json]: Definitely!

**Comment by GrahamCampbell** [comment]
You can't use symfony's dotenv because it doesn't work on php 5.6 or 7.0.

File: lib/Github/Api/Repository/Invitations.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Github\Api\Repository;
+
+use Github\Api\AbstractApi;
+
+class Invitations extends AbstractApi
```
Comment (Nyholm) [inline lib/Github/Api/Repository/Invitations.php]: Please add class docs.

File: lib/Github/Api/Repository/Invitations.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+{
+    public function all($username, $repository, array $params = [])
```
Comment (Nyholm) [inline lib/Github/Api/Repository/Invitations.php]: Please add method docs to all methods in this class. See other APIs/endpoints for example.

File: lib/Github/Api/Repository/Invitations.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    public function decline($invitation)
+    {
+        return $this->delete('/user/repository_invitations/'.rawurlencode($invitation));
+    }
+
+    public function remove($username, $repository, $invitation)
+    {
+        return $this->delete('/repos/'.rawurlencode($username).'/'.rawurlencode($repository).'/invitations/'.rawurlencode($invitation));
```
Comment (Nyholm) [inline lib/Github/Api/Repository/Invitations.php]: Maybe a `sprintf` would improve readability here.

File: lib/Github/Api/Repository/Invitations.php [hunk 10]
```diff
@@ -17,12 +18,24 @@ class Invitations extends AbstractApi
      * @var string
      */
     const READ_PERMISSIONS = 'read';
+    /**
+     * Represents the write permissions.
```
Comment (m1guelpf) [inline lib/Github/Api/Repository/Invitations.php]: Update the description

File: lib/Github/Api/Repository/Invitations.php [hunk 11]
```diff
@@ -17,12 +18,24 @@ class Invitations extends AbstractApi
...
+     */
+    const PULL_PERMISSIONS = 'pull';
     /**
      * Represents the write permissions.
      *
      * @var string
      */
     const WRITE_PERMISSIONS = 'write';
+    /**
+     * Represents the write permissions.
```
Comment (m1guelpf) [inline lib/Github/Api/Repository/Invitations.php]: Update the description

File: test/Github/Tests/Api/TestCase.php [hunk 12]
```diff
@@ -21,7 +21,10 @@ protected function getApiMock()
             ->getMock();
         $httpClient
             ->expects($this->any())
-            ->method('sendRequest');
+            ->method('sendRequest')
+            ->willReturnCallback(function ($a) {
```
Comment (m1guelpf) [inline test/Github/Tests/Api/TestCase.php]: Why is this needed?

**Comment by acrobat** [comment]
I'm also in doubt that we need the integration test for this endpoint. Most (if not all) api classes are covered by unit tests. So maybe in te future we can build a full integration test suite, but for now I think we can leave them.

**Comment by danieldevsquad** [comment]
Those integration tests were useful during development (and might still be). I apologize for all the inconvenience.

**Comment by acrobat** [comment]
No problem @danieldevsquad and your change is still very welcome! So if you are still interested you can reopen your pullrequest!
